### PR TITLE
Fix srv record lookup

### LINKED ISSUE #26116: kube-dns panic
crashlooping on my cluster:

```
I0523 22:11:45.809653       1 dns.go:352] Received DNS Request:galera.cluster.local., exact:false
I0523 22:11:45.809675       1 dns.go:385] records:[], retval:[], path:[local cluster galera]
I0523 22:11:46.064002       1 dns.go:352] Received DNS Request:kubernetes.default.svc.cluster.local., exact:false
I0523 22:11:46.064113       1 dns.go:385] records:[0xc82030e930], retval:[{10.0.0.1 0 10 10  false 30 0  }], path:[local cluster svc default kubernetes]
I0523 22:11:46.064450       1 dns.go:352] Received DNS Request:kubernetes.default.svc.cluster.local., exact:false
I0523 22:11:46.064476       1 dns.go:385] records:[0xc82030e930], retval:[{10.0.0.1 0 10 10  false 30 0  }], path:[local cluster svc default kubernetes]
I0523 22:11:46.064915       1 dns.go:393] Received ReverseRecord Request:1.0.0.10.in-addr.arpa.
I0523 22:11:46.816377       1 dns.go:352] Received DNS Request:galera.e2e-tests-petset-kqc7g.svc.cluster.local., exact:false
I0523 22:11:46.816435       1 dns.go:385] records:[0xc82030f810], retval:[{10.180.3.3 0 10 10  false 30 0  }], path:[local cluster svc e2e-tests-petset-kqc7g galera]
panic: runtime error: slice bounds out of range

goroutine 151 [running]:
panic(0x1677b20, 0xc820014020)
	/usr/local/google/home/abshah/.gvm/gos/go1.6/src/runtime/panic.go:464 +0x3e6
k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/msg.Domain(0x0, 0x0, 0x0, 0x0)
	/usr/local/google/home/abshah/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/msg/service.go:132 +0x2b0
k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server.(*server).SRVRecords(0xc8204601c0, 0xc82011f830, 0x30, 0xe00010021, 0xc82011f830, 0x30, 0x7faaca000200, 0x0, 0x0, 0x0, ...)
	/usr/local/google/home/abshah/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server/server.go:580 +0xfb7
k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server.(*server).ServeDNS(0xc8204601c0, 0x7faaca4e35f0, 0xc8204ead80, 0xc82011d7a0)
	/usr/local/google/home/abshah/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/skynetservices/skydns/server/server.go:392 +0x3558
k8s.io/kubernetes/vendor/github.com/miekg/dns.(*ServeMux).ServeDNS(0xc820464210, 0x7faaca4e35f0, 0xc8204ead80, 0xc82011d7a0)
	/usr/local/google/home/abshah/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:209 +0xbb
k8s.io/kubernetes/vendor/github.com/miekg/dns.(*Server).serve(0xc820452410, 0x7faaca4e20e0, 0xc82011f7d0, 0x7faaca4e20b8, 0xc820464210, 0xc82012ae00, 0x41, 0x200, 0xc820022640, 0xc82012e5a0, ...)
	/usr/local/google/home/abshah/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:576 +0x7ef
created by k8s.io/kubernetes/vendor/github.com/miekg/dns.(*Server).serveUDP
	/usr/local/google/home/abshah/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/miekg/dns/server.go:530 +0x3e1
```

@ArtfulCoder

### PR DESCRIPTION
**Pull request by bprashanth**
Took longer than expected for unrelated reasons. This is intentionally a pretty dumb fix.
Fixes https://github.com/kubernetes/kubernetes/issues/26116

### CODE REVIEW

**Comment by bprashanth** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by bprashanth** [comment]
Failure was the federated-docs.md update that we missed via manual merge

**Comment by ArtfulCoder** [comment]
we might just need to set the key field in our response
https://github.com/openshift/origin/blob/master/pkg/dns/serviceresolver.go#L181

**Comment by bprashanth** [comment]
I actually want it to consistently have the right key, in the tree, instead of appending it from every exit method. But I didn't realize there's a skymsg function I can use to reconstruct the etcd path, that feels safer.

**Comment by bprashanth** [comment]
@k8s-bot unit test this issue: #IGNORE (integration test timeout, logs are useless and we're removing them anyway https://github.com/kubernetes/kubernetes/pull/26016)

**Comment by lukaszo** [comment]
@bprashanth any update on this? I hit this error when testing petsets

**Comment by bprashanth** [comment]
Yeah it'll go in soon, feel free to review it :)
@kubernetes/goog-cluster

File: cmd/kube-dns/dns.go [hunk 1]
```diff
@@ -25,11 +26,18 @@ import (
 	"k8s.io/kubernetes/pkg/version/verflag"
 )
 
+// Global kube-dns verbosity level.
+var logVerbosity = pflag.Bool("verbose-logs", false, "If set, logs are written at V(4), otherwise V(2).")
```
Comment (thockin) [inline cmd/kube-dns/dns.go]: how is this different that just setting --v=4 ?

File: pkg/dns/dns.go [hunk 2]
```diff
@@ -269,17 +285,30 @@ func (kd *KubeDNS) getServiceFromEndpoints(e *kapi.Endpoints) (*kapi.Service, er
 	return nil, fmt.Errorf("got a non service object in services store %v", obj)
 }
 
+// fqdn constructs the fqdn for the given service. subpaths is a list of paths
+// rooted at the given service, ending at a service record.
+func (kd *KubeDNS) fqdn(service *kapi.Service, subpaths ...string) string {
+	domainLabels := append(append(kd.domainPath, serviceSubdomain, service.Namespace, service.Name), subpaths...)
+	reverse := []string{}
```
Comment (thockin) [inline pkg/dns/dns.go]: reverseArray() exists in this pkg

File: pkg/dns/dns.go [hunk 3]
```diff
@@ -269,17 +285,30 @@ func (kd *KubeDNS) getServiceFromEndpoints(e *kapi.Endpoints) (*kapi.Service, er
 	return nil, fmt.Errorf("got a non service object in services store %v", obj)
 }
 
+// fqdn constructs the fqdn for the given service. subpaths is a list of paths
```
Comment (thockin) [inline pkg/dns/dns.go]: s/path/path elements/?

File: pkg/dns/dns.go [hunk 4]
```diff
@@ -269,17 +285,30 @@ func (kd *KubeDNS) getServiceFromEndpoints(e *kapi.Endpoints) (*kapi.Service, er
...
+// rooted at the given service, ending at a service record.
+func (kd *KubeDNS) fqdn(service *kapi.Service, subpaths ...string) string {
+	domainLabels := append(append(kd.domainPath, serviceSubdomain, service.Namespace, service.Name), subpaths...)
+	reverse := []string{}
+	for i := len(domainLabels) - 1; i >= 0; i-- {
+		reverse = append(reverse, domainLabels[i])
+	}
+	return dns.Fqdn(strings.Join(reverse, "."))
```
Comment (thockin) [inline pkg/dns/dns.go]: you could just append a '.' which is all this function does.

File: cmd/kube-dns/dns.go (hunk 1 above)
Comment (bprashanth) [inline cmd/kube-dns/dns.go]: usually you also need to set logtostderr, and there's some other module using --v for something else, so this was getting all related settings and turning verbosity all the way up. Turns out in this case logtostderr defaults to true, so just --v is enough.

File: pkg/dns/dns.go (hunk 4 above)
Comment (bprashanth) [inline pkg/dns/dns.go]: but then I'd have to explain that the dot is appended to construct the fqdn

File: pkg/dns/treecache.go [hunk 5]
```diff
@@ -178,6 +204,20 @@ func (cache *TreeCache) ensureChildNode(path ...string) *TreeCache {
 	return childNode
 }
 
+// printTreeCache prints the given tree cache.
+func printTreeCache(path string, c *TreeCache) {
```
Comment (thockin) [inline pkg/dns/treecache.go]: we have Serialize() already - is that sufficient?

**Comment by lukaszo** [comment]
@bprashanth I reviewed and tested it. It works for me.

**Comment by bprashanth** [comment]
@lukaszo thanks! @thockin queue empty, lgtm?

**Comment by thockin** [comment]
LGTM

**Comment by luxas** [comment]
@bprashanth Seems like you forgot to push the `1.3` images for `arm`, `arm64` and `ppc64le`. 
Please do that ASAP

Right now, I haven't got gcr.io push permission, so can't take care of it.

cc @vishh

**Comment by luxas** [comment]
@girishkalele This ^ is also required for the docker-multinode patch I have

**Comment by bprashanth** [comment]
I pushed arm and arm64, please test. looks like make release doesn't build ARCH=ppc64le, tag me on a pr updating RELEASES.md on how to build them.

**Pull request merged**